source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,591 @@
+        return;
+
+      LinearAlgebra::Vector boundary_velocity;
+
+      const IndexSet &mesh_locally_owned = mesh_deformation_dof_handler.locally_owned_dofs();
+      IndexSet mesh_locally_relevant;
+      DoFTools::extract_locally_relevant_dofs (mesh_deformation_dof_handler,
+                                               mesh_locally_relevant);
+      boundary_velocity.reinit(mesh_locally_owned, mesh_locally_relevant,
+                               this->get_mpi_communicator());
+
+      // Determine the mesh velocity at the surface based on diffusion of
+      // the topography
+      diffuse_boundary(mesh_deformation_dof_handler, mesh_locally_owned,
+                       mesh_locally_relevant, boundary_velocity, boundary_id);
+
+      // now insert the relevant part of the solution into the mesh constraints
+      IndexSet constrained_dofs;
+      DoFTools::extract_boundary_dofs(mesh_deformation_dof_handler,
+                                      ComponentMask(dim, true),
+                                      constrained_dofs,
+                                      boundary_id);
+
+      for (unsigned int i = 0; i < constrained_dofs.n_elements();  ++i)
+        {
+          types::global_dof_index index = constrained_dofs.nth_index_in_set(i);
+          if (mesh_velocity_constraints.can_store_line(index))
+            if (mesh_velocity_constraints.is_constrained(index)==false)
+              {
+                mesh_velocity_constraints.add_line(index);
+                mesh_velocity_constraints.set_inhomogeneity(index, boundary_velocity[index]);
+              }
+        }
+    }
+
+
+    template <int dim>
+    void Diffusion<dim>::declare_parameters(ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh deformation");
+      {
+        prm.enter_subsection("Diffusion");
+        {
+          prm.declare_entry("Hillslope transport coefficient", "1e-6",
+                            Patterns::Double(0),
+                            "The hillslope transport coefficient $\\kappa$ used to "
+                            "diffuse the free surface, either as a  "
+                            "stabilization step or a to mimic erosional "
+                            "and depositional processes. Units: $\\text{m}^2/\\text{s}$. ");
+          prm.declare_entry("Time between diffusion", "1e20",
+                            Patterns::Double(0,std::numeric_limits<double>::max()),
+                            "The time between each application of diffusion. "
+                            "Units: years if the "
+                            "'Use years in output instead of seconds' parameter is set; "
+                            "seconds otherwise.");
+          prm.declare_entry("Time steps between diffusion", "1",
+                            Patterns::Integer(0,std::numeric_limits<int>::max()),
+                            "The number of time steps between each application of "
+                            "diffusion.");
```

Comment:
Can you document what happens if one specifies both `Time between diffusion` and `Time steps between diffusion`? 

Maybe separately: Why do you actually need both? Is this so expensive that you wouldn't want to do it in every time step? And if so, why not just use it at fixed intervals in time (rather than in time steps)?

Reply:
I think this question is not yet resolved? If you only want to diffuse every n timesteps, because you do not want to smooth too much, could you not instead reduce the diffusion coefficient?

Reply:
Wolfgang's question is resolved after we talked about it in person: I don't need both and removed `Time between diffusion`.